<a href="https://colab.research.google.com/github/FadQode/Data_analysis/blob/main/DS3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [183]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Prequisite: Defining Analysis Goal

Menjawab Pertanyaan berikut yang mungkin bisa bermanfaat dan memberi Insight pada departemen HR untuk menentukan langkah terbaik yang perlu diambil perusahaan berikutnya

## Defining Bussiness Question

### Operational Performance
1. Seberapa efektif dan efisien tim PIC HR dalam menangani dan menyelesaikan permintaan layanan karyawan dari waktu ke waktu sesuai target SLA yang ditetapkan?

### Root Cause & Prioritization

2. Apa faktor utama penyebab keterlambatan penyelesaian request HR, dan bagaimana prioritas penanganannya dapat ditentukan berdasarkan jenis permintaan dan PIC yang bertanggung jawab

# Prequisite: Dataset Definition

## Column Definition and Completness

### Read Dataset

In [184]:
df = pd.read_csv('/content/HR_DATA.csv')

### Asses Data Type and Missing Values

In [185]:
df = pd.read_csv("/content/HR_DATA.csv", delimiter=';')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   RequestID        1000 non-null   object
 1   EmpID            1000 non-null   int64 
 2   EmpName          1000 non-null   object
 3   Dept             1000 non-null   object
 4   Request_Date     991 non-null    object
 5   Completion_Date  907 non-null    object
 6   Request_Detail   994 non-null    object
 7   PIC              995 non-null    object
dtypes: int64(1), object(7)
memory usage: 62.6+ KB


In [186]:
df.isna().sum()

RequestID           0
EmpID               0
EmpName             0
Dept                0
Request_Date        9
Completion_Date    93
Request_Detail      6
PIC                 5
dtype: int64

In [187]:
df.head()

RequestID  EmpID        EmpName Dept Request_Date Completion_Date  \
0   REQ0001    103  Charlie Davis   HR   01/01/2024      09/01/2024   
1   REQ0002    103  Charlie Davis   HR   01/01/2024      02/01/2024   
2   REQ0003    105     Eve Wilson   HR   01/01/2024      03/01/2024   
3   REQ0004    104    David Brown   HR   02/01/2024      10/01/2024   
4   REQ0005    102      Bob Smith   HR   02/01/2024      04/01/2024   

    Request_Detail               PIC  
0    Payroll Issue  Sarah Supervisor  
1  Access Approval      John Manager  
2  Access Approval      John Manager  
3    Payroll Issue  Sarah Supervisor  
4    Benefit Claim      John Manager

Proses ini menunjukkan bahwa terdapat ketidaksesuaian dalam dataset berupa hala hal berikut:

1. Perbedaan tipe data antara RequestID dan EMP ID yang keduanya sama sama menujukkan ID sehingga lebih baik memiliki tipe data yang sama
2. Request dan Completion Date yang masih berupa object dan bukan DateTime
3. Beberapa kolom memiliki missing values terutama dalam Request Date, Completion_Date, dan Request_Detail


## Column Types Adjustment

In [188]:
# Normalisasi format tanggal menjadi DD/MM/YYYY
df["Request_Date"] = pd.to_datetime(df["Request_Date"], format="%d/%m/%Y", errors='coerce')
df["Completion_Date"] = pd.to_datetime(df["Completion_Date"], format="%d/%m/%Y", errors='coerce')

# Cek hasil normalisasi
print(df[["Request_Date", "Completion_Date"]].head())

  Request_Date Completion_Date
0   2024-01-01      2024-01-09
1   2024-01-01      2024-01-02
2   2024-01-01      2024-01-03
3   2024-01-02      2024-01-10
4   2024-01-02      2024-01-04


In [189]:
df['EmpID'] = df['EmpID'].astype(str)
df['RequestID'] = df['RequestID'].astype(str)

In [190]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   RequestID        1000 non-null   object        
 1   EmpID            1000 non-null   object        
 2   EmpName          1000 non-null   object        
 3   Dept             1000 non-null   object        
 4   Request_Date     991 non-null    datetime64[ns]
 5   Completion_Date  907 non-null    datetime64[ns]
 6   Request_Detail   994 non-null    object        
 7   PIC              995 non-null    object        
dtypes: datetime64[ns](2), object(6)
memory usage: 62.6+ KB


1. Untuk Kolom dan FItur ID di konsistensikan menjadi tipe data STR untuk keduanya supaya konsisten
2. Menyamakan format Tanggal untuk Request & Completion Date, serta mengubah tipe datanya menjadi DateTime

# 1. Feature Engineering

### Menmisahkan bulan bulan Penanganan request dari Tanggal untuk memudahkan analisis dan viuslisasi nantinya

In [191]:
df['Request_Month'] = df['Request_Date'].dt.to_period('M')
df['Completion_Month'] = df['Completion_Date'].dt.to_period('M')

## Defining Inflow

### Nilai Value Request dari tanggal dengan Request Terbanyak

In [192]:
df['Request_Date'].value_counts().sort_values(ascending=False)

Request_Date
2024-09-02    10
2024-06-20     9
2024-09-03     8
2024-03-07     8
2024-12-06     7
              ..
2024-04-24     1
2024-04-06     1
2024-04-15     1
2024-12-07     1
2024-12-13     1
Name: count, Length: 338, dtype: int64

### Mengelompokkan nilai Request Inflow dari Bulan dengan Request Berdasarkan Bulan

In [193]:
inflow = df.groupby('Request_Month').size()

In [194]:
inflow

Request_Month
2024-01    74
2024-02    70
2024-03    86
2024-04    78
2024-05    82
2024-06    89
2024-07    78
2024-08    76
2024-09    88
2024-10    94
2024-11    91
2024-12    85
Freq: M, dtype: int64

In [195]:
inflow.describe()

count    12.000000
mean     82.583333
std       7.427937
min      70.000000
25%      77.500000
50%      83.500000
75%      88.250000
max      94.000000
dtype: float64

In [196]:
inflow.sum()

np.int64(991)

Inflow sendiri adalah Request yang masuk untuk diproses atau ditindak lanjuti oleh PIC.
Berikut adalah insight yang didapat dari pemrosesan diatas
1. Terdapat 991 nilai masuk dari total 1000 berarti kemungkinan terdapat 9 data yang memiliki nilai null pada Request_Date sehingga terdapat 9 unresolved case
2. Request tertinggi datang pada bulan November 2024 dengan total 94 permintaan, sedangkan request terendah datang pada bulan Februari 2024
3. Rata rata sebanyak 83 permintaan masuk setiap bulannya

## Defining Outflow

### Nilai Value Request dari tanggal dengan Request Terbanyak

In [197]:
df['Completion_Date'].value_counts().sort_values(ascending=False)

Completion_Date
2024-11-26    13
2024-06-21    10
2024-05-20     8
2024-03-10     7
2024-12-17     6
              ..
2024-03-28     1
2024-03-14     1
2024-12-07     1
2025-01-09     1
2025-01-03     1
Name: count, Length: 339, dtype: int64

### Mengelompokkan nilai Completion Outflow Berdasarkan Bulan

In [198]:
outflow = df.groupby('Completion_Month').size()

In [199]:
outflow

Completion_Month
2024-01    54
2024-02    65
2024-03    73
2024-04    72
2024-05    76
2024-06    82
2024-07    78
2024-08    67
2024-09    73
2024-10    87
2024-11    86
2024-12    84
2025-01    10
Freq: M, dtype: int64

In [200]:
outflow.describe()

count    13.000000
mean     69.769231
std      20.199645
min      10.000000
25%      67.000000
50%      73.000000
75%      82.000000
max      87.000000
dtype: float64

In [201]:
outflow.sum()

np.int64(907)

Outflow menghitung jumlah perminataan(request) yang telah diselesaikan, dari completion outflow diatas berikut insisht yang didapatkan:
1. hanya terdapat 907 Completion dibanding jumlah data yang 1000 ataupun jumlah inflow yang 991 yang berarti terdapat 93 data yang belum terselesaikan
2. rata rata penyelesaian request perbulannya adalah 80 lebih rendah jika dibandingkan rerata request masuk perbulannya yang ada di angka 83
3. bulan paling produktif dalam menyelesaikan request adalah pada November 2024, jumlah yang sama dengan permintaan terbanyak, dan paling tidak produktif pada bulan Januari 2025 kemungkinan menyelesaikan open requests yang terdapat pada bula bulan sebelumnya

## Defining Backlog

### Jumlah Unresolved Case per Januari 2025

In [202]:
backlog = df[(df['Request_Date'].notna()) & (df['Completion_Date'].isna())]
print(f"Jumlah backlog (unresolved cases): {backlog.shape[0]}")

Jumlah backlog (unresolved cases): 93


### Tren Unresolved Case per bulannya

In [203]:
monthly_backlog = df[(df['Request_Date'].notna()) & (df['Completion_Date'].isna())]
monthly_backlog['Request_Month'] = monthly_backlog['Request_Date'].dt.to_period('M').astype(str)
backlog_trend = monthly_backlog.groupby('Request_Month').size()

<ipython-input-203-dc4278e58173>:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [204]:
backlog_trend

Request_Month
2024-01     8
2024-02     9
2024-03    10
2024-04     4
2024-05     6
2024-06     6
2024-07     5
2024-08     9
2024-09     8
2024-10    10
2024-11     9
2024-12     9
dtype: int64

Backlog adalah total unresolved case, permintaan yang masuk tetapi belum terselesaikan, berikut insight yang didapat dari proses mendefinisikan backlog:
1. terdapat 907 kasus yang sudah terselesaikan dan terdapat 93 kasus yang unresolved case, memiliki star date tapi tidak completion date, dibuktikan dari jumlah backlog
2. jumlah backlog terbanyak terdapat pada bulan November 2024 dan maret 2024

## Defining Open Requests

In [205]:
open_requests = df[df['Request_Date'].isna()].shape[0]
print(f"Jumlah open requests (belum dimulai): {open_requests}")

Jumlah open requests (belum dimulai): 9


In [206]:
monthly_open = df[df['Request_Date'].isna()]
monthly_open['Completion_Month'] = monthly_open['Completion_Date'].dt.to_period('M').astype(str)
open_trend = monthly_open.groupby('Completion_Month').size()

<ipython-input-206-625da07ac03a>:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [207]:
open_trend

Completion_Month
2024-10    1
2024-11    5
2024-12    2
2025-01    1
dtype: int64

Open requests adalah Permintaan yang belum memiliki Request Date, artinya belum dimulai oleh tim HR atau belum secara resmi tercatat kapan dimulai.

Insight yang didapat saat mendefinisikab open request:
1. terdapat 9 request yang sudah terselesaikan namun tidak ada tanggal mulai nya, jumlah yang sama dengan jumlah nilai yang hilang pada inflow
2. terjadi hanya pada 5 bulan terakhir
3. terjadi hingga 5 kali pada bulan November 2025, bulan dengan request tertinggi, bisa jadi saat beban kerja tinggi membuat administrasi data jadi kurang teliti atau bisa saja ada migrasi sistem atau human error pada periode itu.

## Open Requests & Backlog Trend Visualization

In [208]:
import plotly.graph_objects as go

fig = go.Figure()

# Backlog line
fig.add_trace(go.Scatter(x=backlog_trend.index, y=backlog_trend.values,
                         mode='lines+markers', name='Backlog', line=dict(color='red')))

# Open Requests line
fig.add_trace(go.Scatter(x=open_trend.index, y=open_trend.values,
                         mode='lines+markers', name='Open Requests', line=dict(color='orange')))

fig.update_layout(title='Monthly Backlog & Open Requests Trend',
                  xaxis_title='Month',
                  yaxis_title='Number of Requests',
                  xaxis_tickangle=-45)
fig.show()

Backlog Trend (garis merah):
Fluktuasi terlihat sepanjang tahun 2024

- Januari hingga Maret: backlog naik, mencapai puncaknya di Maret 2024.
- April: terjadi penurunan signifikan, kemungkinan besar ada upaya penyelesaian backlog secara besar-besaran.
- Juli–Desember: backlog kembali naik dan cenderung stabil di angka tinggi (8–10) → bisa mengindikasikan kapasitas penyelesaian yang tidak cukup cepat untuk menurunkan backlog.
- Backlog tetap tinggi hingga akhir tahun
- Desember menunjukkan backlog belum terselesaikan secara signifikan.

Bisa menandakan bahwa jumlah inflow tetap tinggi dan melebihi kapasitas penyelesaian (outflow), atau prioritas penanganan belum merata

Open Requests Trend (garis oranye):
- Open request hanya muncul di akhir tahun
- Mulai terlihat dari Oktober 2024 hingga Januari 2025., Ini menunjukkan munculnya masalah pencatatan Request Date baru-baru ini.
- Terjadi lonjakan di November (5 kasus) → perlu investigasi apakah ada masalah sistem atau input human error.
- Open request menurun di Januari 2025
- Ini bisa positif (karena mulai ditangani), atau hanya belum bertambah data.



# 2. Handling missing categorical values

## Memilah dan Mengecek niai categorical

In [209]:
cat_col = df.select_dtypes(include='object')

In [210]:
cat_col.isna().sum()

RequestID         0
EmpID             0
EmpName           0
Dept              0
Request_Detail    6
PIC               5
dtype: int64

In [211]:
df[cat_col.isna().any(axis=1)]

RequestID EmpID        EmpName Dept Request_Date Completion_Date  \
19    REQ0020   102      Bob Smith   HR   2024-01-12      2024-01-19   
98    REQ0099   105     Eve Wilson   HR   2024-02-10      2024-02-17   
265   REQ0266   101  Alice Johnson   HR   2024-04-14      2024-04-21   
320   REQ0321   104    David Brown   HR   2024-05-04      2024-05-08   
470   REQ0471   103  Charlie Davis   HR   2024-06-27      2024-07-04   
674   REQ0675   103  Charlie Davis   HR   2024-09-11      2024-09-12   
751   REQ0752   103  Charlie Davis   HR   2024-10-10      2024-10-19   
810   REQ0811   103  Charlie Davis   HR   2024-10-30      2024-11-05   
819   REQ0820   102      Bob Smith   HR   2024-11-01      2024-11-11   
841   REQ0842   101  Alice Johnson   HR   2024-11-09      2024-11-10   
964   REQ0965   105     Eve Wilson   HR   2024-12-16      2024-12-25   

          Request_Detail               PIC Request_Month Completion_Month  
19         Benefit Claim               NaN       2024-01          2024-01  
98   ID Card Replacement               NaN       2024-02          2024-02  
265                  NaN      John Manager       2024-04          2024-04  
320        Payroll Issue               NaN       2024-05          2024-05  
470  ID Card Replacement               NaN       2024-06          2024-07  
674      Access Approval               NaN       2024-09          2024-09  
751                  NaN      John Manager       2024-10          2024-10  
810                  NaN  Sarah Supervisor       2024-10          2024-11  
819                  NaN      John Manager       2024-11          2024-11  
841                  NaN  Sarah Supervisor       2024-11          2024-11  
964                  NaN      Michael Lead       2024-12          2024-12

In [212]:
missing_counts = df[['PIC', 'Request_Detail']].isna().sum()
missing_percentage = df[['PIC', 'Request_Detail']].isna().mean() * 100

print("Jumlah Missing Values:")
print(missing_counts)
print("\nPersentase Missing Values:")
print(missing_percentage)

Jumlah Missing Values:
PIC               5
Request_Detail    6
dtype: int64

Persentase Missing Values:
PIC               0.5
Request_Detail    0.6
dtype: float64


Tahapan yang dilakukan:
1. Memisahkan categorical columns dengan dataset lainnya
2. mengecek data data dengan categorical features yang kosong
3. mengecek presentase nilai kosong

insight yang didapat:
1. missing values hanya terdapat pada kolom Request_Detail dan PIC
2. setiap missing value untuk PIC dan request_detail tidak ada yang beririsan(semua terdapat pada data yang berbeda)
3. hanya 0.5 persen data PIC yang hilang dan 0.6 persen data request_detail, jika dijumlahkan hanya 1.1 persen data yang hilang sehingga hanya sangat sedikit data yang hilang

## Missing Values Handling

In [213]:
df['PIC'].fillna('Unknown PIC', inplace=True)
df['Request_Detail'].fillna('Not Specified', inplace=True)

<ipython-input-213-891cf3c8fdf2>:1: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



<ipython-input-213-891cf3c8fdf2>:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or d

In [214]:
# df['PIC_Missing'] = df['PIC'].isna().astype(int)
# df['Request_Detail_Missing'] = df['Request_Detail'].isna().astype(int)

# df['PIC'].fillna(df['PIC'].mode()[0], inplace=True)
# df['Request_Detail'].fillna(df['Request_Detail'].mode()[0], inplace=True)

# df.head()

Tahapan yang dilakukan:
1. karena hanya sebagian kecil seluruh data kategorikal yang hilang akan diganti dengan Unknown untuk kolom PIC yang kosong dan not specified untuk req detail yang kosong
2. selanjutnya akan dianalisa lebih lanjut mengenai data data kosong ini

# 3. SLA Compliance Analysis

In [215]:
holidays = [
    "2024-01-01", "2024-02-08", "2024-02-10", "2024-02-14", "2024-03-11", "2024-03-29",
    "2024-03-31", "2024-04-10", "2024-04-11", "2024-05-01", "2024-05-09", "2024-05-23",
    "2024-06-01", "2024-06-17", "2024-07-07", "2024-08-17", "2024-09-16", "2024-12-25",
    "2024-02-09", "2024-03-12", "2024-04-08", "2024-04-09", "2024-04-12", "2024-04-15",
    "2024-05-10", "2024-05-24", "2024-06-18", "2024-12-26"
]



In [216]:
holidays = np.array(pd.to_datetime(holidays).date, dtype='datetime64[D]')

def calculate_processing_days(row):
    request_date = row['Request_Date']
    completion_date = row['Completion_Date']

    if pd.isna(request_date) or pd.isna(completion_date):
        return np.nan

    processing_days = np.busday_count(
        request_date.date(),
        completion_date.date(),
        holidays=holidays
    )

    return processing_days

In [217]:
df['Processing_Days'] = df.apply(calculate_processing_days, axis=1)

# Step 3: Buat kolom SLA Compliance
df['SLA_Compliance'] = np.where(df['Processing_Days'] <= 3, 'Pass', 'Miss')

# Tampilkan hasil
df[['Request_Date', 'Completion_Date', 'Processing_Days', 'SLA_Compliance']].head()


Request_Date Completion_Date  Processing_Days SLA_Compliance
0   2024-01-01      2024-01-09              5.0           Miss
1   2024-01-01      2024-01-02              0.0           Pass
2   2024-01-01      2024-01-03              1.0           Pass
3   2024-01-02      2024-01-10              6.0           Miss
4   2024-01-02      2024-01-04              2.0           Pass

In [218]:
# Hitung jumlah 'Pass' dan 'Miss'
sla_summary = df['SLA_Compliance'].value_counts()

# Tampilkan hasil
print("Perbandingan SLA Compliance:")
print(sla_summary)

# Jika ingin dalam bentuk persentase juga:
sla_percentage = df['SLA_Compliance'].value_counts(normalize=True) * 100
print("\nPersentase SLA Compliance:")
print(sla_percentage)

Perbandingan SLA Compliance:
SLA_Compliance
Miss    560
Pass    440
Name: count, dtype: int64

Persentase SLA Compliance:
SLA_Compliance
Miss    56.0
Pass    44.0
Name: proportion, dtype: float64


Tahapan yang dilakukan:
1. Mendefinisikan Hari libur yang terdapat pada tahun 2024 di Indonesia
2. Mendfinisikan fungsi yang dapat menghitung hari pemrosesan dan memetakan Request Date dan Completion date ke dalam SLA
3. menjalankan fungsi terhadap dataframe dan menandai dengan SLA pass jika pemrosesan kurang dari sama dengan 3 hari bisnis dan miss jika lebih dari 3 hari bisnis
4. Membuat Quick Summary untuk membandingkan jumlah Pass dan Miss pada SLA

Insight:
1. Jumlah dan presentase Miss lebih banyak daripada Pass
2. Mengingat Jumlah Inflow yang juga lebih tinggi dibanding outflow serta lebih banyak SLA miss dibandingkan pass, bisa jadi para karyawan dan pegawai HR kewalahan dalam memeberikan layanan

# 4. Performance Insights

In [219]:
import plotly.express as px
import plotly.graph_objects as go

print(f"Inflow: {inflow.sum()}")
print(f"Outflow: {outflow.sum()}")
print(f"Backlog: {backlog_trend.sum()}")
print(f"Open Requests: {open_trend.sum()}")

Inflow: 991
Outflow: 907
Backlog: 93
Open Requests: 9


In [220]:
data_combined = pd.DataFrame({
    'Month': inflow.index.astype(str),
    'Inflow': inflow.values,
    'Outflow': outflow.reindex(inflow.index, fill_value=0).values
})


fig = go.Figure()
fig.add_trace(go.Scatter(x=data_combined['Month'], y=data_combined['Inflow'], mode='lines+markers', name='Inflow'))
fig.add_trace(go.Scatter(x=data_combined['Month'], y=data_combined['Outflow'], mode='lines+markers', name='Outflow'))
fig.add_trace(go.Scatter(x=backlog_trend.index, y=backlog_trend.values,mode='lines+markers', name='Backlog', line=dict(color='orange')))
fig.update_layout(title='Inflow, Outflow, Backlog', xaxis_title='Month', yaxis_title='Count')
fig.show()

# Step 4: Stacked Bar Chart
fig_bar = go.Figure()
fig_bar.add_trace(go.Bar(x=data_combined['Month'], y=data_combined['Inflow'], name='Inflow'))
fig_bar.add_trace(go.Bar(x=data_combined['Month'], y=data_combined['Outflow'], name='Outflow'))
fig_bar.update_layout(barmode='group', title='Stacked Bar Chart: Inflow vs Outflow', xaxis_title='Month', yaxis_title='Count')
fig_bar.show()

In [221]:
# # prompt: tampilkan req dan completion date outflow januari

# # Filter data for January outflow
# january_outflow = df[df['Completion_Month'] == '2025-01']

# # Display Request and Completion Dates for January outflow
# print(january_outflow[['Request_Date', 'Completion_Date', 'Outflow', "Inflow"]])


Langkah-Langkah yang dilakukan:
1. Gabungkan Inflow dan Outflow ke satu dataframe Terstruktur untuk memudahkan visualisasi
2. Tambahkan Figure untuk line chart dan tampilkan trend data inflow, outflow, dan backlog
3. Tambahkan figure untuk bar chart untuk melihat perbandingan secara langsung tren inflow dibanding outflow per bulannya

Insight:
1. Surplus Permintaan (Inflow > Outflow)
    - Terlihat hampir di seluruh bulan (kecuali Juli 2024), jumlah permintaan yang masuk lebih banyak dibandingkan yang berhasil diselesaikan. Ini berpotensi menyebabkan penumpukan backlog
2. Penurunan Backlog di April 2024
    - Penurunan tajam backlog di bulan April 2024 menandakan bahwa ada peningkatan efisiensi tim dalam menangani permintaan.
3. Stabilitas Backlog
    - Setelah April, backlog cenderung naik sedikit tapi stabil (antara 9–10),
    - Namun, tidak ada penurunan signifikan setelah Agustus—hal ini perlu menjadi perhatian agar backlog tidak terus stagnan.
4. Keseimbangan Inflow dan Outflow (Juli 2024)
    - Bulan Juli menjadi satu-satunya momen di mana Outflow > Inflow.
5. Persebaran Backlog
    - Jumlah backlog yang mencapai 93 cenderung tersebar secara merata pada seluruh bulan yang kemungkinan menjadi penyebab surplus permintaan di seluruh bulan (kecuali Juli)

# 5. SLA Compliance Trend

In [222]:
sla_compliance = df.groupby([df['Request_Date'].dt.to_period('M'), 'SLA_Compliance']).size().unstack(fill_value=0)
sla_compliance

SLA_Compliance  Miss  Pass
Request_Date              
2024-01           48    26
2024-02           34    36
2024-03           39    47
2024-04           33    45
2024-05           37    45
2024-06           42    47
2024-07           37    41
2024-08           47    29
2024-09           55    33
2024-10           67    27
2024-11           54    37
2024-12           58    27

In [223]:
import plotly.graph_objects as go

fig_sla = go.Figure()
fig_sla.add_trace(go.Bar(
    x=sla_compliance.index.astype(str),
    y=sla_compliance.get('Pass', 0),
    name='Pass',
    marker_color='green'
))
fig_sla.add_trace(go.Bar(
    x=sla_compliance.index.astype(str),
    y=sla_compliance.get('Miss', 0),
    name='Miss',
    marker_color='red'
))

fig_sla.update_layout(
    barmode='group',
    title='SLA Compliance Trend per Month (Grouped)',
    xaxis_title='Month',
    yaxis_title='Count'
)

fig_sla.show()

Langkah:
1. Melihat Persebaran Miss dan Pass Perbulannya
2. Melakukan visulisasi menggunakan bar chart untuk memudahkan analisa dan komparasi

Insight:
1. Persebaran Pass dan Miss
    - Terlihat Request yang pass unggul hanya di sekitar Quartal 1 dan Quartal 2 kecuali bulan januari ditambah bulan juli
    - Pada Kuartal 3 dan 4 (Jul–Des 2024), jumlah request yang tidak memenuhi SLA (Miss) terlihat jauh lebih tinggi dibandingkan yang memenuhi (Pass).
2. Indikasi Beban Berlebih atau Bottleneck:
    - Ketimpangan yang tinggi antara Miss dan Pass kemungkinan disebabkan oleh:
    - Kapasitas penyelesaian tidak sebanding dengan volume permintaan.
    - Adanya kendala operasional (misalnya SDM terbatas, proses lambat, atau gangguan sistem).
    - Tidak adanya penyesuaian resource meskipun inflow stabil/tetap tinggi.

# 6. Backlog Management

In [224]:
pre_datefilled_df = df.copy()

backlog_requests = pre_datefilled_df[pre_datefilled_df['Completion_Date'].isna()]
display(backlog_requests)

RequestID EmpID        EmpName Dept Request_Date Completion_Date  \
5     REQ0006   105     Eve Wilson   HR   2024-01-03             NaT   
23    REQ0024   102      Bob Smith   HR   2024-01-14             NaT   
38    REQ0039   103  Charlie Davis   HR   2024-01-18             NaT   
57    REQ0058   102      Bob Smith   HR   2024-01-25             NaT   
63    REQ0064   104    David Brown   HR   2024-01-27             NaT   
..        ...   ...            ...  ...          ...             ...   
983   REQ0984   101  Alice Johnson   HR   2024-12-24             NaT   
989   REQ0990   102      Bob Smith   HR   2024-12-27             NaT   
991   REQ0992   105     Eve Wilson   HR   2024-12-27             NaT   
996   REQ0997   103  Charlie Davis   HR   2024-12-30             NaT   
999   REQ1000   103  Charlie Davis   HR   2024-12-31             NaT   

          Request_Detail               PIC Request_Month Completion_Month  \
5          Benefit Claim      John Manager       2024-01              NaT   
23       Access Approval      Michael Lead       2024-01              NaT   
38   ID Card Replacement      John Manager       2024-01              NaT   
57       Access Approval      Michael Lead       2024-01              NaT   
63       Access Approval      John Manager       2024-01              NaT   
..                   ...               ...           ...              ...   
983  ID Card Replacement  Sarah Supervisor       2024-12              NaT   
989        Payroll Issue      John Manager       2024-12              NaT   
991  ID Card Replacement  Sarah Supervisor       2024-12              NaT   
996  ID Card Replacement      Michael Lead       2024-12              NaT   
999        Benefit Claim      John Manager       2024-12              NaT   

     Processing_Days SLA_Compliance  
5                NaN           Miss  
23               NaN           Miss  
38               NaN           Miss  
57               NaN           Miss  
63               NaN           Miss  
..               ...            ...  
983              NaN           Miss  
989              NaN           Miss  
991              NaN           Miss  
996              NaN           Miss  
999              NaN           Miss  

[93 rows x 12 columns]

In [225]:
df_pre_imputed = df.copy()

In [226]:
df_fill_completion_only = df_pre_imputed.copy()
median_processing_time = df_fill_completion_only['Processing_Days'].median()

df_fill_completion_only.loc[df_fill_completion_only['Completion_Date'].isna(), 'Completion_Date'] = (
    df_fill_completion_only['Request_Date'] + pd.to_timedelta(median_processing_time, unit='D')
)
df.loc[df['Completion_Date'].isna(), 'Completion_Date'] = df_fill_completion_only['Completion_Date']


In [227]:
df_fill_request_only = df_pre_imputed.copy()
mean_start_delay = (df_fill_request_only['Completion_Date'] - df_fill_request_only['Request_Date']).dt.days.median()

df_fill_request_only.loc[df_fill_request_only['Request_Date'].isna(), 'Request_Date'] = (
    df_fill_request_only['Completion_Date'] - pd.to_timedelta(mean_start_delay, unit='D')
)

df.loc[df['Request_Date'].isna(), 'Request_Date'] = df_fill_request_only['Request_Date']


In [228]:
df['Processing_Days'] = (df['Completion_Date'] - df['Request_Date']).dt.days
backlog_requests['Processing_Days'] = median_processing_time
backlog_requests['SLA_Compliance'] = backlog_requests['Processing_Days'].apply(lambda x: 'Pass' if x <= 3 else 'Miss')

df['Request_Month'] = df['Request_Date'].dt.to_period('M')
df['Completion_Month'] = df['Completion_Date'].dt.to_period('M')


df[['Request_Date', 'Completion_Date', 'Processing_Days', 'SLA_Compliance', 'Request_Month', "Completion_Month"]].head()

<ipython-input-228-8afe92ab5f26>:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-228-8afe92ab5f26>:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Request_Date Completion_Date  Processing_Days SLA_Compliance Request_Month  \
0   2024-01-01      2024-01-09                8           Miss       2024-01   
1   2024-01-01      2024-01-02                1           Pass       2024-01   
2   2024-01-01      2024-01-03                2           Pass       2024-01   
3   2024-01-02      2024-01-10                8           Miss       2024-01   
4   2024-01-02      2024-01-04                2           Pass       2024-01   

  Completion_Month  
0          2024-01  
1          2024-01  
2          2024-01  
3          2024-01  
4          2024-01

In [229]:
df.isna().sum()

RequestID           0
EmpID               0
EmpName             0
Dept                0
Request_Date        0
Completion_Date     0
Request_Detail      0
PIC                 0
Request_Month       0
Completion_Month    0
Processing_Days     0
SLA_Compliance      0
dtype: int64

Langkah-Langkah:
1. Salin Data Asli
  - Sebelum mengubah apapun, kita simpan salinan dataframe asli untuk jaga-jaga.
2. Identifikasi Backlog (Completion Date Kosong)
  - Kita cari request yang belum selesai, yaitu yang Completion_Date-nya masih kosong.
3. Isi Completion Date yang Kosong
  - Hitung median dari waktu pemrosesan (Processing_Days), lalu gunakan itu untuk mengisi nilai Completion_Date yang kosong.
  - Pendekatan ini mengasumsikan bahwa waktu proses request yang belum selesai serupa dengan request lainnya secara umum (median), sehingga hasilnya lebih realistis daripada sekadar isi dengan tanggal tertentu.
4. Mengisi Nilai Kosong pada Request_Date
  - Hitung median durasi dari Request_Date ke Completion_Date. Kurangkan durasi ini dari Completion_Date untuk memperkirakan Request_Date.
  - Pengisian ini dilakukan dengan logika terbalik dari waktu proses, karena kita tahu kapan request selesai dan mengasumsikan waktu masuknya mirip dengan kasus-kasus lainnya.





# 7. PIC Performance Analysis

In [230]:
df.groupby(['PIC']).size().reset_index(name='Count')

PIC  Count
0      John Manager    331
1      Michael Lead    326
2  Sarah Supervisor    338
3       Unknown PIC      5

In [231]:
# PIC Performance Analysis
pic_backlog = backlog_requests.groupby(['PIC', 'Request_Detail']).size().reset_index(name='Count')

# Enhanced PIC performance analysis with all requests (not just backlog)
pic_all_requests = df.groupby(['PIC', 'Request_Detail']).size().reset_index(name='Total_Count')
fig_all_req = px.bar(pic_all_requests, x='PIC', y='Total_Count', color='Request_Detail', title='Total Requests per PIC (Including Completed)', barmode='stack')
fig_all_req.show()

In [232]:

unknown_pic_count = df[df['PIC'] == 'Unknown PIC'].shape[0]
print(f"Number of data with 'Unknown PIC': {unknown_pic_count}")


Number of data with 'Unknown PIC': 5


In [233]:
# Pie chart per PIC
fig_pic = px.pie(pic_backlog, names='PIC', values='Count', title='Backlog Requests per PIC')
fig_pic.show()

# Pie Chart: Percentage of Requests Completed per PIC
pic_completed = df[~df['Completion_Date'].isna()].groupby('PIC').size().reset_index(name='Completed_Count')
pic_total = df.groupby('PIC').size().reset_index(name='Total_Count')
pic_summary = pic_total.merge(pic_completed, on='PIC', how='left').fillna(0)
pic_summary['Completion_Percentage'] = (pic_summary['Completed_Count'] / pic_summary['Total_Count']) * 100

fig_completion = px.pie(
    pic_summary,
    names='PIC',
    values='Completed_Count',
    title='Number of Requests Completed per PIC'
)
fig_completion.show()

# Additional Pie Charts for Each Request Category
PIC = df['PIC'].unique()
for category in PIC:
    category_data = df[df['PIC'] == category].groupby(['Request_Detail']).size().reset_index(name='Count')
    fig_category = px.pie(category_data, names='Request_Detail', values='Count', title=f'PIC {category} Solved Precentafge')
    fig_category.show()


In [234]:

# Additional Pie Charts for Each Request Category
request_categories = df['Request_Detail'].unique()
for category in request_categories:
    category_data = df[df['Request_Detail'] == category].groupby(['PIC']).size().reset_index(name='Count')
    fig_category = px.pie(category_data, names='PIC', values='Count', title=f'Requests for {category} per PIC')
    fig_category.show()

Langkah:

1. Mencari tahu Jumlah Request yang telah diselesaikan per PIC
2. Visualisasi Stack Chart penyelesaian Request per PIC
3. Visualisasi Pie Chart penyelesaian Request per PIC
4. Visualisasi Pie chart presentase penyelesaian PIC pada suatu Request

Insight:
1. Kinerja PIC
Ketiga PIC menunjukkan performa kerja yang serupa dan solid.
  - Sarah menjadi yang paling produktif dengan total 338 request yang berhasil diselesaikan.
  - Disusul oleh Michael dengan 331 request, dan
  - John dengan 326 request
2. Terdapat 5 request yang tidak memiliki informasi terkait PIC yang bertanggung jawab.
3. Jenis Request Tidak Terspesifikasi
  - Masing-masing PIC sempat menangani beberapa request yang tidak disebutkan jenisnya (Request_Detail kosong atau NULL).
4. Secara umum, baik dari sisi jumlah backlog maupun jenis request yang diselesaikan, distribusi tugas cenderung merata di antara ketiga PIC.

# 8. Targeted Table

In [261]:
# Filter request yang sebelumnya tidak punya Completion_Date
predicted_completion_df = df_fill_completion_only[
    df_pre_imputed['Completion_Date'].isna()
][['Request_Date', 'Completion_Date']]

predicted_completion_df = predicted_completion_df.rename(columns={
    'Completion_Date': 'Predicted_Completion_Date'
})

display(predicted_completion_df)


Request_Date Predicted_Completion_Date
5     2024-01-03                2024-01-07
23    2024-01-14                2024-01-18
38    2024-01-18                2024-01-22
57    2024-01-25                2024-01-29
63    2024-01-27                2024-01-31
..           ...                       ...
983   2024-12-24                2024-12-28
989   2024-12-27                2024-12-31
991   2024-12-27                2024-12-31
996   2024-12-30                2025-01-03
999   2024-12-31                2025-01-04

[93 rows x 2 columns]

In [262]:
# Filter request yang sebelumnya tidak punya Request_Date
predicted_request_df = df_fill_request_only[
    df_pre_imputed['Request_Date'].isna()
][['Request_Date', 'Completion_Date']]

predicted_request_df = predicted_request_df.rename(columns={
    'Request_Date': 'Predicted_Request_Date'
})

display(predicted_request_df)


Predicted_Request_Date Completion_Date
785             2024-10-18      2024-10-23
802             2024-10-29      2024-11-03
821             2024-11-04      2024-11-09
846             2024-11-08      2024-11-13
869             2024-11-21      2024-11-26
896             2024-11-21      2024-11-26
920             2024-11-30      2024-12-05
953             2024-12-11      2024-12-16
995             2025-01-04      2025-01-09

In [235]:
predicted_dates = df[['RequestID', 'Request_Date', 'Completion_Date']]
predicted_dates.head()

RequestID Request_Date Completion_Date
0   REQ0001   2024-01-01      2024-01-09
1   REQ0002   2024-01-01      2024-01-02
2   REQ0003   2024-01-01      2024-01-03
3   REQ0004   2024-01-02      2024-01-10
4   REQ0005   2024-01-02      2024-01-04

In [240]:
import pandas as pd
import plotly.graph_objects as go

inflow_before = pre_datefilled_df .groupby('Request_Month').size()
outflow_before = pre_datefilled_df .groupby('Completion_Month').size()


inflow_after = df.groupby('Request_Month').size()
outflow_after = df.groupby('Completion_Month').size()


# Gabungkan seluruh bulan dari inflow dan outflow
all_months_before = inflow_before.index.union(outflow_before.index)
all_months_after = inflow_after.index.union(outflow_after.index)

# Reindex agar sinkron
inflow_before = inflow_before.reindex(all_months_before, fill_value=0)
outflow_before = outflow_before.reindex(all_months_before, fill_value=0)

inflow_after = inflow_after.reindex(all_months_after, fill_value=0)
outflow_after = outflow_after.reindex(all_months_after, fill_value=0)

# Buat dataframe
data_combined_before = pd.DataFrame({
    'Month': all_months_before.astype(str),
    'Inflow': inflow_before.values,
    'Outflow': outflow_before.values
})

data_combined_after = pd.DataFrame({
    'Month': all_months_after.astype(str),
    'Inflow': inflow_after.values,
    'Outflow': outflow_after.values
})

# Ubah indeks period menjadi string untuk visualisasi
inflow_before.index = inflow_before.index.astype(str)
outflow_before.index = outflow_before.index.astype(str)


inflow_after.index = inflow_after.index.astype(str)
outflow_after.index = outflow_after.index.astype(str)


# Plot Inflow
fig1 = go.Figure()
fig1.add_trace(go.Scatter(x=inflow_before.index, y=inflow_before.values, name='Inflow Before', line=dict(color='blue', dash='dot')))
fig1.add_trace(go.Scatter(x=inflow_after.index, y=inflow_after.values, name='Inflow After', line=dict(color='blue')))
fig1.update_layout(title='Monthly Inflow (Before vs After)', xaxis_title='Month', yaxis_title='Number of Requests')
fig1.show()

# Plot Outflow
fig2 = go.Figure()
fig2.add_trace(go.Scatter(x=outflow_before.index, y=outflow_before.values, name='Outflow Before', line=dict(color='green', dash='dot')))
fig2.add_trace(go.Scatter(x=outflow_after.index, y=outflow_after.values, name='Outflow After', line=dict(color='green')))
fig2.update_layout(title='Monthly Outflow (Before vs After)', xaxis_title='Month', yaxis_title='Number of Requests')
fig2.show()





In [253]:
import plotly.graph_objects as go

# 1. Filter backlog dari data sebelum dan sesudah imputasi
backlog_before = pre_datefilled_df[(pre_datefilled_df['Request_Date'].notna()) & (pre_datefilled_df['Completion_Date'].isna())]
backlog_after = df[(df['Request_Date'].notna()) & (df['Completion_Date'].isna())]

# 2. Tampilkan jumlah backlog total
print(f"Jumlah backlog BEFORE (unresolved cases): {backlog_before.shape[0]}")
print(f"Jumlah backlog AFTER (unresolved cases): {backlog_after.shape[0]}")

# 3. Hitung backlog trend per bulan
trend_backlog_before = backlog_before.groupby(backlog_before['Request_Date'].dt.to_period('M')).size()
trend_backlog_after = backlog_after.groupby(backlog_after['Request_Date'].dt.to_period('M')).size()

# 4. Ubah ke string agar mudah divisualisasikan
trend_backlog_before.index = trend_backlog_before.index.astype(str)
trend_backlog_after.index = trend_backlog_after.index.astype(str)

# 5. Gabungkan semua bulan agar sumbu X sinkron
all_months = trend_backlog_before.index.union(trend_backlog_after.index)

trend_backlog_before = trend_backlog_before.reindex(all_months, fill_value=0)
trend_backlog_after = trend_backlog_after.reindex(all_months, fill_value=0)

# 6. Plot backlog trend per bulan
fig = go.Figure()
fig.add_trace(go.Scatter(x=trend_backlog_before.index, y=trend_backlog_before.values,
                         name='Backlog Before', line=dict(color='orange', dash='dot')))
fig.add_trace(go.Scatter(x=trend_backlog_after.index, y=trend_backlog_after.values,
                         name='Backlog After', line=dict(color='orange')))
fig.update_layout(title='Monthly Backlog Trend (Before vs After)',
                  xaxis_title='Month', yaxis_title='Number of Unresolved Requests')
fig.show()


Jumlah backlog BEFORE (unresolved cases): 93
Jumlah backlog AFTER (unresolved cases): 0


In [260]:
import plotly.graph_objects as go

# ---------------------
# Chart 1: BEFORE
fig_bar_before = go.Figure()
fig_bar_before.add_trace(go.Bar(x=inflow_before.index, y=inflow_before.values,
                                name='Inflow'))  # Default blue
fig_bar_before.add_trace(go.Bar(x=outflow_before.index, y=outflow_before.values,
                                name='Outflow'))  # Default orange

fig_bar_before.update_layout(
    barmode='group',
    title='Monthly Inflow vs Outflow (Before Filling)',
    xaxis_title='Month',
    yaxis_title='Number of Requests',
    legend_title='Category',
    bargap=0.2
)
fig_bar_before.show()

# ---------------------
# Chart 2: AFTER
fig_bar_after = go.Figure()
fig_bar_after.add_trace(go.Bar(x=inflow_after.index, y=inflow_after.values,
                               name='Inflow'))  # Default blue
fig_bar_after.add_trace(go.Bar(x=outflow_after.index, y=outflow_after.values,
                               name='Outflow'))  # Default orange

fig_bar_after.update_layout(
    barmode='group',
    title='Monthly Inflow vs Outflow (After Filling)',
    xaxis_title='Month',
    yaxis_title='Number of Requests',
    legend_title='Category',
    bargap=0.2
)
fig_bar_after.show()


In [242]:
# Step 1: Menghitung Inflow, Outflow, Backlog, Open Requests
df['Request_Month'] = df['Request_Date'].dt.to_period('M')
df['Completion_Month'] = df['Completion_Date'].dt.to_period('M')

inflow = df.groupby('Request_Month').size()
outflow = df.groupby('Completion_Month').size()
backlog = inflow.cumsum() - outflow.cumsum()
backlog[backlog < 0] = 0  # Pastikan backlog tidak negatif
open_requests = inflow.sum() - outflow.sum()

# Step 2: Menampilkan Summary Cards
print(f"Inflow: {inflow.sum()}")
print(f"Outflow: {outflow.sum()}")
print(f"Backlog: {backlog.iloc[-1]}")
print(f"Open Requests: {open_requests}")


Inflow: 1000
Outflow: 1000
Backlog: 0
Open Requests: 0


# 9. Advanced Analysis - Defining another Backlog

In [243]:
monthly_backlog = inflow - outflow

monthly_backlog

Request_Month
2024-01    15
2024-02    -7
2024-03     4
2024-04     1
2024-05     0
2024-06     1
2024-07    -3
2024-08    -2
2024-09     8
2024-10     0
2024-11    -1
2024-12    -5
2025-01   -11
Freq: M, dtype: int64

In [244]:
data_combined['Backlog Change'] = data_combined['Inflow'] - data_combined['Outflow']
data_combined['Cumulative Backlog'] = data_combined['Backlog Change'].cumsum()

fig = go.Figure()
fig.add_trace(go.Scatter(x=data_combined['Month'], y=data_combined['Backlog Change'], mode='lines+markers', name='Backlog Change', line=dict(color='orange')))
fig.add_trace(go.Scatter(x=data_combined['Month'], y=data_combined['Cumulative Backlog'], mode='lines+markers', name='Cumulative Backlog', line=dict(color='red', dash='dot')))
fig.add_trace(go.Scatter(x=backlog_trend.index, y=backlog_trend.values,mode='lines+markers', name='Backlog', line=dict(color='green')))
fig.show()

In [245]:
# Step 5: Waterfall Chart
data_waterfall = []
current_backlog = 0
for i, row in data_combined.iterrows():
    net_change = row['Backlog Change']
    current_backlog += net_change
    data_waterfall.append({'Month': row['Month'], 'Backlog Change': net_change})

waterfall_df = pd.DataFrame(data_waterfall)
fig_waterfall = px.bar(waterfall_df, x='Month', y='Backlog Change', title='Waterfall Chart: Backlog Changes', labels={'Backlog Change': 'Change in Backlog'}, text_auto=True)
fig_waterfall.show()


In [252]:
data_combined_after['Backlog Change'] = data_combined_after['Inflow'] - data_combined_after['Outflow']
data_combined_before['Backlog Change'] = data_combined_before['Inflow'] - data_combined_before['Outflow']


backlog_change_after = data_combined_after['Backlog Change']
backlog_change_before = data_combined_before['Backlog Change']

backlog_change_after.index = backlog_change_after.index.astype(str)
backlog_change_before.index = backlog_change_before.index.astype(str)

# Plot Backlog
fig3 = go.Figure()
fig3.add_trace(go.Scatter(x=backlog_change_before.index, y=backlog_change_before.values, name='Backlog Before', line=dict(color='red', dash='dot')))
fig3.add_trace(go.Scatter(x=backlog_change_after.index, y=backlog_change_after.values, name='Backlog After', line=dict(color='red')))
fig3.add_trace(go.Scatter(x=backlog_change_before.index, y=backlog_change_before.values, name='Backlog Before', line=dict(color='red', dash='dot')))
fig3.update_layout(title='Monthly Backlog (Before vs After)', xaxis_title='Month', yaxis_title='Number of Requests')
fig3.show()

In [248]:
# Waterfall Chart: Backlog Change Before
data_waterfall_before = []
for i, row in data_combined_before.iterrows():
    net_change = row['Backlog Change']
    month = row['Month']
    data_waterfall_before.append({'Month': month, 'Backlog Change': net_change})

waterfall_df_before = pd.DataFrame(data_waterfall_before)
fig_waterfall_before = px.bar(
    waterfall_df_before,
    x='Month',
    y='Backlog Change',
    title='Waterfall Chart: Backlog Changes (Before)',
    labels={'Backlog Change': 'Change in Backlog'},
    text_auto=True,
    color='Backlog Change',

)
fig_waterfall_before.update_layout(coloraxis_showscale=False)
fig_waterfall_before.show()


# Waterfall Chart: Backlog Change After
data_waterfall_after = []
for i, row in data_combined_after.iterrows():
    net_change = row['Backlog Change']
    month = row['Month']
    data_waterfall_after.append({'Month': month, 'Backlog Change': net_change})

waterfall_df_after = pd.DataFrame(data_waterfall_after)
fig_waterfall_after = px.bar(
    waterfall_df_after,
    x='Month',
    y='Backlog Change',
    title='Waterfall Chart: Backlog Changes (After)',
    labels={'Backlog Change': 'Change in Backlog'},
    text_auto=True,
    color='Backlog Change',

)
fig_waterfall_after.update_layout(coloraxis_showscale=False)
fig_waterfall_after.show()
